soundcard/utils.py

```diff
@@ -1,21 +1,39 @@
+import re
+
 def match_device(id, devices):
     """Find id in a list of devices.
 
-    id can be a platfom-specific id, a substring of the device name, or a
+    id can be a platfom specific id, a substring of the device name, or a
     fuzzy-matched pattern for the microphone name.
-
     """
     devices_by_id = {device.id: device for device in devices}
-    devices_by_name = {device.name: device for device in devices}
+    real_devices_by_name = {
+        device.name: device for device in devices
+        if not getattr(device, 'isloopback', True)}
+    loopback_devices_by_name = {
+        device.name: device for device in devices
+        if getattr(device, 'isloopback', True)}
     if id in devices_by_id:
         return devices_by_id[id]
+    for device_map in real_devices_by_name, loopback_devices_by_name:
+        if id in device_map:
+            return device_map[id]
     # try substring match:
-    for name, device in devices_by_name.items():
-        if id in name:
-            return device
+    for device_map in real_devices_by_name, loopback_devices_by_name:
+        for name, device in device_map.items():
+            if id in name:
+                return device
     # try fuzzy match:
-    pattern = '.*'.join(id)
-    for name, device in devices_by_name.items():
-        if re.match(pattern, name):
-            return device
+    id_parts = list(id)
+    # Escape symbols in the provided id that have a special meaning
+    # in regular expression to prevent syntax errors e.g. for
+    # unbalanced parentheses.
+    for special_re_char in r'.^$*+?{}\[]|()':
```

Comment:
Very good addition, thank you!

Reply:
Argh, right. I completely forgot about that function… Used now.